src/modules/kvs/kvs.c

```diff
@@ -2914,15 +2914,28 @@ static int checkpoint_get (flux_t *h, char *buf, size_t len, int *seq)
  */
 static int checkpoint_put (flux_t *h, const char *rootref, int rootseq)
 {
-    flux_future_t *f = NULL;
+    flux_future_t *f1 = NULL;
+    flux_future_t *f2 = NULL;
     int rv = -1;
 
-    if (!(f = kvs_checkpoint_commit (h, NULL, rootref, rootseq, 0, 0))
-        || flux_rpc_get (f, NULL) < 0)
+    /* first must ensure all content is flushed */
+    if (!(f1 = flux_rpc (h, "content.flush", NULL, 0, 0))
+        || flux_rpc_get (f1, NULL) < 0) {
+        /* fallthrough to kvs_checkpoint_commit(), ENOSYS may be due
+         * to no backing store, but checkpoint can still be done to
+         * content cache.
+         */
+        if (errno != ENOSYS)
+            goto error;
+    }
+
+    if (!(f2 = kvs_checkpoint_commit (h, NULL, rootref, rootseq, 0, 0))
+        || flux_rpc_get (f2, NULL) < 0)
         goto error;
```

Comment:
Why fall through to the checkpoint commit?  If there is no backing store then wouldn't it also fail?

Reply:
this admittedly surprised me, but you can checkpoint to the content cache without it going through to the backing store.

Reply:
ya know, thinking about this a bit .... do we want that behavior?  Do we want checkpoints to work with the content cache and w/o a backing store.  It seems like a nuanced corner case.  I'm not entirely sure why it was initially supported that way.  This would be for a different issue I think, as this PR solves a specific `content.flush` before "checkpoint" corner case.

Reply:
If there is a hook in the content cache for the checkpoint, why don't we do the cache flush there?

Reply:
Sorry I've forgotten so many details here!

Reply:
hmmm.  supporting `checkpoint` w/o a backing module simply may have been a fallout from supporting the `none` backing module.  https://github.com/flux-framework/flux-core/pull/4492     perhaps checkpoint specifically should not be supported by the content cache.

Reply:
I guess it allows the kvs module to be reloaded and not lose all the data when there is no backing store?

Reply:
> I guess it allows the kvs module to be reloaded and not lose all the data when there is no backing store?

I do think that's the point.  But that is sort of a nuanced use case, possibly only useful in testing?

Reply:
> If there is a hook in the content cache for the checkpoint, why don't we do the cache flush there?

That's not a bad idea.  Lemme look into that, although it'd be for a different PR b/c `FLUX_KVS_SYNC` does two calls and that would have to be collapsed down into one as well.